coderd/httpmw/apikey_test.go

```diff
@@ -991,3 +991,237 @@ func TestAPIKey(t *testing.T) {
 		require.Equal(t, http.StatusOK, res.StatusCode)
 	})
 }
+
+func TestAPIKeyExpiryRefresh(t *testing.T) {
+	t.Parallel()
+
+	t.Run("ShortLivedKeyRefresh", func(t *testing.T) {
+		t.Parallel()
+		var (
+			db, _ = dbtestutil.NewDB(t)
+			user  = dbgen.User(t, db, database.User{})
+			now   = time.Now()
+			// 2 minute lifetime
+			lifetime = int64(120)
+			// Key expires in 30 seconds (less than half the 2-minute lifetime)
+			sentAPIKey, token = dbgen.APIKey(t, db, database.APIKey{
+				UserID:          user.ID,
+				LastUsed:        dbtime.Now().Add(-time.Hour),
+				ExpiresAt:       dbtime.Time(now.Add(30 * time.Second)),
+				LifetimeSeconds: lifetime,
+			})
+			r  = httptest.NewRequest("GET", "/", nil)
+			rw = httptest.NewRecorder()
+		)
+		r.Header.Set(codersdk.SessionTokenHeader, token)
+
+		// Mock time to be exactly now
+		successHandler := http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			httpapi.Write(r.Context(), rw, http.StatusOK, codersdk.Response{
+				Message: "It worked!",
+			})
+		})
+
+		httpmw.ExtractAPIKeyMW(httpmw.ExtractAPIKeyConfig{
+			DB:              db,
+			RedirectToLogin: false,
+		})(successHandler).ServeHTTP(rw, r)
+		res := rw.Result()
+		defer res.Body.Close()
+		require.Equal(t, http.StatusOK, res.StatusCode)
+
+		// Verify the key was refreshed
+		gotAPIKey, err := db.GetAPIKeyByID(r.Context(), sentAPIKey.ID)
+		require.NoError(t, err)
+
+		// For a 2-minute key, refresh threshold is 1 minute (half the lifetime)
+		// Since the key expires in 30 seconds and threshold is 1 minute,
+		// it should be refreshed
+		require.True(t, gotAPIKey.ExpiresAt.After(sentAPIKey.ExpiresAt),
+			"API key should have been refreshed for short-lived key")
+		// Should be extended by the full lifetime (2 minutes)
+		expectedExpiry := now.Add(time.Duration(lifetime) * time.Second)
+		require.WithinDuration(t, expectedExpiry, gotAPIKey.ExpiresAt, time.Second,
+			"API key should be extended by full lifetime")
+	})
+
+	t.Run("ShortLivedKeyNoRefresh", func(t *testing.T) {
```

Comment:
I feel like these would all make more sense in a test table, the only difference is the lifetime.

Also feel like maybe it can use a range.